### DISCUSSION TITLE
Accessibility of protected members of attribute classes

### DISCUSSION THREAD

**Discussion by abelbraaksma** [dotnet/csharplang#3948] [General]
From the current spec, sections on accessibility and attributes do not mention how protected members of an attribute and protected attributes are scoped w.r.t. attribute usages.

Furthermore, it is unclear or at least under-specified, whether protected members of classes are accessible from within an attribute declaration of an inheriting class. In fact, the spec appears to prohibit such use as it [talks about **program text** of a type](https://github.com/dotnet/csharplang/blob/master/spec/basic-concepts.md#accessibility-domains):

> and the program text of a type is defined as all program text contained in the _type_declarations_ of that type (including, possibly, types that are nested within the type

Here, `type_declarations` are within the type and do not include the class declaration itself, afaik.

### Example #&#x200b;1

This example is taken from this discussion: https://github.com/dotnet/fsharp/issues/3172.

```c#
public class TestClassBase
{
  protected const string Foo = "Bar";
}
```

The following is valid as per the status quo, but appears to be invalid per the spec:

```c#
// protected const is accessible, though spec appears to disallow it
[Trait(Foo, "")]
public class MyCSharpTestClass : TestClassBase
{
}
```

### Example #&#x200b;2
```c#
public class MyAttribute: Attribute
{
    protected const int TestField = 42;
    protected MyAttribute(int x)
    {

    }
}
```

The following is valid, though again the spec appears to prohibit this, in particular [by this statement in the spec](https://github.com/dotnet/csharplang/blob/master/spec/attributes.md#attribute-instances):

> If C does not have public accessibility, then a compile-time error occurs.

```c#
// protected ctor does not throw error, but spec disallows this
// and: protected const is accessible outside its scope
[My(MyAttribute.TestField)]
public class OtherAttribute: MyAttribute
{
    public OtherAttribute(int x): base(x) { }
}
```

### Example #&#x200b;3 

If I add a `protected int TestProp;` then suddenly the rules of the spec seem to apply properly and indeed, it is not accessible. Which suggests that the protected constant above has crept in the status quo, but does not align with the spec. This rule [is actually in the spec](https://github.com/dotnet/csharplang/blob/master/spec/attributes.md#run-time-retrieval-of-an-attribute-instance):

> Let Name be the identifier of the named_argument Arg. If Name does not identify a non-static public read-write field or property on O, then an exception is thrown.

(the spec has this as a runtime error, which is somewhat surprising...)

![image](https://user-images.githubusercontent.com/16015770/94367283-ff8af000-00dd-11eb-96ce-a618abed964b.png)

----

I do not propose to suddenly start prohibiting this, but I do think we can do better in the section an accessibility to include whether or not the attribute declaration of a type or member are part of the _program text of the type_ or not, and in the section on attributes, what we mean by _public_, as clearly attributes can be used when they are protected as well.

**Comment by 333fred** [comment]
It's very likely that the rules the C# compiler implements here do not match the spec exactly, due do the native compiler's behavior. Name lookup especially was a place where the native compiler often played a bit loose with the rules, and in these cases we often matched Roslyn's behavior to the native compiler in order to allow people to move forward. I do agree that a reading of the spec suggests these cases should be illegal, but that's the likely reason for the behavior.

**Reply by abelbraaksma** [reply]
Thanks @333fred. I do think, though, that we ought to say something in the spec by clarifying the extend to where "program text" applies. Likewise for the unclear distinction between runtime and compile time errors of the attribute constructor (when not public). 

Or we could change it such that the "playing loose" rules of the C# compiler become the _de jure_ rules.

**Comment by abelbraaksma** [comment]
Time to close this :)
